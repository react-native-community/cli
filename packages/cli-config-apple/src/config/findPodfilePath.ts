/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import {inlineString, logger} from '@react-native-community/cli-tools';
import path from 'path';
import findAllPodfilePaths from './findAllPodfilePaths';
import {ApplePlatform} from '../types';
import {supportedPlatforms} from './supportedPlatforms';

// Regexp matching all test projects
const TEST_PROJECTS = /test|example|sample/i;

// Podfile in the bundle package
const BUNDLE_VENDORED_PODFILE = 'vendor/bundle/ruby';

export default function findPodfilePath(
  cwd: string,
  platformName: ApplePlatform,
) {
  const podfiles = findAllPodfilePaths(cwd)
    /**
     * Then, we will run a simple test to rule out most example projects,
     * unless they are located in a `platformName` folder
     */
    .filter((project) => {
      if (path.dirname(project) === platformName) {
        // Pick the Podfile in the default project (in the iOS folder)
        return true;
      }

      if (TEST_PROJECTS.test(project)) {
        // Ignore the Podfile in test and example projects
        return false;
      }

      if (project.indexOf(BUNDLE_VENDORED_PODFILE) > -1) {
        // Ignore the podfile shipped with Cocoapods in bundle
        return false;
      }

      // Accept all the others
      return true;
    })
    /**
     * Podfile from `platformName` folder will be picked up as a first one.
     */
    .sort((project) => (path.dirname(project) === platformName ? -1 : 1));

  const supportedPlatformsArray: string[] = Object.values(supportedPlatforms);
  const containsUnsupportedPodfiles = podfiles.every(
    (podfile) => !supportedPlatformsArray.includes(podfile.split('/')[0]),
  );

  if (podfiles.length > 0) {
    if (podfiles.length > 1 && containsUnsupportedPodfiles) {
      logger.warn(
        inlineString(`
          Multiple Podfiles were found: ${podfiles}. Choosing ${podfiles[0]} automatically.
          If you would like to select a different one, you can configure it via "project.${platformName}.sourceDir".
          You can learn more about it here: https://github.com/react-native-community/cli/blob/main/docs/configuration.md
        `),
      );
    }
    return path.join(cwd, podfiles[0]);
  }

  return null;
}
