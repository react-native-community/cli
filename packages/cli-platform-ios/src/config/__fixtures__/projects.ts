// @ts-ignore FIXME: Add global types for jest
const path = jest.requireActual('path');
// @ts-ignore
const fs = jest.requireActual('fs');

export const projectWithPodfileOnly = {
  ios: {},
};

export const project = {
  ios: {
    Podfile: 'content',
    'demoProject.xcodeproj': {
      'project.pbxproj': fs.readFileSync(
        path.join(__dirname, './files/project.pbxproj'),
      ),
    },
  },
};

export const withExamples = {
  ...project,
  Examples: {
    Podfile: 'content',
    'exampleProject.xcodeproj': {
      'project.pbxproj': fs.readFileSync(
        path.join(__dirname, './files/project.pbxproj'),
      ),
    },
  },
};
