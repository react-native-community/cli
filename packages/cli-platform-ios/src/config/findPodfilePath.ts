/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import {inlineString, logger} from '@react-native-community/cli-tools';
import path from 'path';
import findAllPodfilePaths from './findAllPodfilePaths';

// Regexp matching all test projects
const TEST_PROJECTS = /test|example|sample/i;

// Base iOS folder
const IOS_BASE = 'ios';

export default function findPodfilePath(cwd: string) {
  const podfiles = findAllPodfilePaths(cwd)
    /**
     * Then, we will run a simple test to rule out most example projects,
     * unless they are located in a `ios` folder
     */
    .filter(
      (project) =>
        path.dirname(project) === IOS_BASE || !TEST_PROJECTS.test(project),
    )
    /**
     * Podfile from `ios` folder will be picked up as a first one.
     */
    .sort((project) => (path.dirname(project) === IOS_BASE ? -1 : 1));

  if (podfiles.length > 0) {
    if (podfiles.length > 1) {
      logger.warn(
        inlineString(`
          Multiple Podfiles were found: ${podfiles}. Choosing ${podfiles[0]} automatically.
          If you would like to select a different one, you can configure it via "project.ios.sourceDir".
          You can learn more about it here: https://github.com/react-native-community/cli/blob/master/docs/configuration.md
        `),
      );
    }
    return path.join(cwd, podfiles[0]);
  }

  return null;
}
