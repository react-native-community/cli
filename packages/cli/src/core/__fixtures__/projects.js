/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import android from './android';
import ios from './ios';

const flat = {
  android: android.valid,
  ios: ios.valid,
  Podfile: 'empty',
};

const nested = {
  android: {
    app: android.valid,
  },
  ios: ios.valid,
};

const withExamples = {
  Examples: flat,
  ios: ios.valid,
  android: android.valid,
};

const withPods = {
  Podfile: 'content',
  ios: ios.pod,
};

module.exports = { flat, nested, withExamples, withPods };
