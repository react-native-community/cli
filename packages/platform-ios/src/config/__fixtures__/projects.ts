// @ts-ignore FIXME: Add global types for jest
const path = jest.requireActual('path');
// @ts-ignore
const fs = jest.requireActual('fs');

const ios = {
  'demoProject.xcodeproj': {
    'project.pbxproj': fs.readFileSync(
      path.join(__dirname, './files/project.pbxproj'),
    ),
  },
};

const iosPod = {
  'demoProject.xcodeproj': {
    'project.pbxproj': fs.readFileSync(
      path.join(__dirname, './files/project.pbxproj'),
    ),
  },
  'TestPod.podspec': 'empty',
};

export const flat = {
  ios,
};

export const nested = {
  ios,
};

export const withExamples = {
  Examples: flat,
  ios,
};

export const withPods = {
  Podfile: 'content',
  ios: iosPod,
};

export const withoutPods = {
  ios,
};
