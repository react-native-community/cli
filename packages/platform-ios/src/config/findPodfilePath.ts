/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import {inlineString, logger} from '@react-native-community/cli-tools';
import glob from 'glob';
import path from 'path';

// Regexp matching all test projects
const TEST_PROJECTS = /test|example|sample/i;

// Base iOS folder
const IOS_BASE = 'ios';

// These folders will be excluded from search to speed it up
const GLOB_EXCLUDE_PATTERN = ['**/@(Pods|node_modules|Carthage)/**'];

export default function findPodfilePath(cwd: string) {
  /**
   * First, we're going to look for all Podfiles within the `cwd`
   */
  const podfiles = glob
    .sync('**/Podfile', {
      cwd,
      ignore: GLOB_EXCLUDE_PATTERN,
    })
    /**
     * Then, we will run a simple test to rule out most example projects,
     * unless they are located in a `ios` folder
     */
    .filter(
      (project) =>
        path.dirname(project) === IOS_BASE || !TEST_PROJECTS.test(project),
    )
    /**
     * Podfile from `ios` folder will be picked up as a first one.
     */
    .sort((project) => (path.dirname(project) === IOS_BASE ? -1 : 1));

  if (podfiles.length > 0) {
    if (podfiles.length > 1) {
      logger.warn(
        inlineString(`
          Multiple Podfiles were found: ${podfiles}. Choosing ${podfiles[0]} automatically.
          If you would like to select a different one, you can configure it via "project.ios.sourceDir".
          You can learn more about it here: https://github.com/react-native-community/cli/blob/master/docs/configuration.md
        `),
      );
    }
    return path.join(cwd, podfiles[0]);
  }

  return null;
}
