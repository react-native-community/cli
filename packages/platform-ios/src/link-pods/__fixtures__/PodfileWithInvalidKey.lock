PODS:
  - MyPackage (1.0.0)

EXTERNAL SOURCES:
  ExternalInterface:
    :path: !ruby/object:Pathname
    path: "../node_modules/MyOtherPackage/ios"

SPEC CHECKSUMS:
  MyPackage: 77fd5fb102a4a5eedafa2c2b0245ceb7b7c15e45
  MyOtherPackage: a9bb76128853e98a9ef6d12b0c8c91debc9bc475


PODFILE CHECKSUM: a8110dc7c367fc529b8b6a1084258784444d62ec

COCOAPODS: 1.7.5
