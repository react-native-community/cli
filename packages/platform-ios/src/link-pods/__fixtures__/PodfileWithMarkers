source 'https://github.com/CocoaPods/Specs.git'
# platform :ios, '9.0'

target 'None' do
  # Uncomment the next line if you're using Swift or would like to use dynamic frameworks
  # use_frameworks!
  # Your 'node_modules' directory is probably in the root of your project, # but if not, adjust the `:path` accordingly 
  pod 'React', :path => '../node_modules/react-native', :subspecs => [
    'Core',
    'RCTText',
    'RCTNetwork',
    'BatchedBridge',
    'RCTImage',
    'RCTWebSocket', # needed for debugging 
    # Add any other subspecs you want to use in your project 
  ]
  
  # Add new pods below this line  

  # test should point to this line
  target 'NoneTests' do
    inherit! :search_paths
    # Pods for testing
  end
end

target 'Second' do
  
  target 'NoneUITests' do
    inherit! :search_paths
    # Add new pods below this line
  end

end