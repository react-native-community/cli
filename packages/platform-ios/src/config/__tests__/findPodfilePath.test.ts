import findPodfilePath from '../findPodfilePath';
import * as projects from '../__fixtures__/projects';

jest.mock('path');
jest.mock('fs');

const fs = require('fs');

describe('ios::findPodfilePath', () => {
  it('returns null if there is no Podfile', () => {
    fs.__setMockFilesystem(projects.withoutPods);
    expect(findPodfilePath(process.cwd(), '')).toBeNull();
  });

  it('returns Podfile path if it exists', () => {
    fs.__setMockFilesystem(projects.withPods);
    expect(findPodfilePath(process.cwd(), '/ios')).toContain('Podfile');
  });
});
