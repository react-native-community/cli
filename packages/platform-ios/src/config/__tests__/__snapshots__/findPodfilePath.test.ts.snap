// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ios::findPodfilePath prints a warning when multile Podfiles are found 1`] = `
Array [
  Array [
    "Multiple Podfiles were found: bar/ios/Podfile,foo/ios/Podfile. Choosing bar/ios/Podfile automatically. If you would like to select a different one, you can configure it via \\"project.ios.sourceDir\\". You can learn more about it here: https://github.com/react-native-community/cli/blob/master/docs/configuration.md",
  ],
]
`;
